linalg/dtensor.hpp

```diff
@@ -0,0 +1,172 @@
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_DTENSOR
+#define MFEM_DTENSOR
+
+#include "../general/okina.hpp"
+
+namespace mfem
+{
+
+// Simple vector and tensor classes for dense linear algebra on the device
+
+/// Class for a simple Vector of sizes, appropriate for use on the GPU
+class DeviceVector3
+{
+private:
+   double data[3];
+public:
+   DeviceVector3() {}
+   DeviceVector3(const double *r) { data[0]=r[0]; data[1]=r[1]; data[2]=r[2]; }
+   DeviceVector3(const double r0, const double r1, const double r2 = 0.0)
+   { data[0]=r0; data[1]=r1; data[2]=r2; }
+   inline operator double* () { return data; }
+   inline operator const double* () const { return data; }
+   inline double& operator[](const size_t i) { return data[i]; }
+   inline double& operator()(const int i) { return data[i]; }
+   inline const double& operator()(const int i) const { return data[i]; }
+};
+
+/// A Class to compute the real index from the multi-indices of a tensor
+template <int N, int Dim, typename T, typename... Args>
+class TensorInd
+{
+public:
+   MFEM_HOST_DEVICE static int result(const int* sizes, T first, Args... args)
+   {
+#ifndef MFEM_USE_CUDA
+      MFEM_ASSERT(first<sizes[N-1],"Trying to access out of boundary.");
+#endif
+      return first + sizes[N - 1] * TensorInd < N + 1, Dim, Args... >::result(sizes,
+                                                                              args...);
+   }
+};
+// Terminal case
+template <int Dim, typename T, typename... Args>
+class TensorInd<Dim, Dim, T, Args...>
+{
+public:
+   MFEM_HOST_DEVICE static int result(const int* sizes, T first, Args... args)
+   {
+#ifndef MFEM_USE_CUDA
+      MFEM_ASSERT(first<sizes[Dim-1],"Trying to access out of boundary.");
+#endif
+      return first;
+   }
+};
+
+/// A class to initialize the size of a Tensor
+template <int N, int Dim, typename T, typename... Args>
+class Init
+{
+public:
+   static int result(int* sizes, T first, Args... args)
+   {
+      sizes[N - 1] = first;
+      return first * Init < N + 1, Dim, Args... >::result(sizes, args...);
+   }
+};
+// Terminal case
+template <int Dim, typename T, typename... Args>
+class Init<Dim, Dim, T, Args...>
+{
+public:
+   static int result(int* sizes, T first, Args... args)
+   {
+      sizes[Dim - 1] = first;
+      return first;
+   }
+};
+
+/// A basic generic Tensor class, appropriate for use on the GPU
+template<int Dim, typename Scalar = double>
+class DeviceTensor
+{
+
+protected:
+   int capacity;
+   Scalar *data;
+   int sizes[Dim];
```

Comment:
Does it make sense to precompute and store the strides for each dimension as well?

Reply:
I don't think this makes a difference, the strides are inlined so the difference should be minimal.

Reply:
I suppose it will add a few multiplications for each indexing operation. Maybe this matters for very high-dimensional tensors? In most cases I think you're right that the difference should be minimal.

Reply:
The integer units on NVIDIA GPUs are quite weak (far weaker than the FPU).  I think it would be worth our while to store the strides and avoid unnecessary mults for each index.